File: Leetcode Daily Challenge/024-June-2025/008-lexicographical-numbers.java
// https://leetcode.com/problems/lexicographical-numbers/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(1)
    public List<Integer> lexicalOrder(int n) {
        List<Integer> result = new ArrayList<>();
        for (int i = 1; i <= Math.min(n, 9); i ++) {
            result.add(i);
            solve(i, n, result);
        }
        return result;
    }
    private void solve(int val, int n, List<Integer> result) {
        for (int i = 0; i <= 9; i ++) {
            int nextVal = (val * 10 + i);
            if (nextVal <= n) {
                result.add(nextVal);
                solve(nextVal, n, result);
            } else break;
        }
    }
}

File: Leetcode Daily Challenge/024-June-2025/008-lexicographical-numbers.cpp
// https://leetcode.com/problems/lexicographical-numbers/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    void solve(vector<int> &result, long val, int n) {
        for (int i = 0; i <= 9; i ++) {
            int next = val * 10 + i;
            if (next <= n) {
                result.push_back(next);
                solve(result, next, n);
            } else break;
        }
    }
public:
    // TC : O(N)
    // SC : O(1)
    vector<int> lexicalOrder(int n) {
        vector<int> result;
        for (int i = 1; i <= min(n, 9); i ++) {
            result.push_back(i);
            solve(result, i, n);
        }
        return result;
    }
};